### ISSUE TITLE
cockroachdb compatibility

### ISSUE THREAD

**Issue by markbates** [markbates/pop#82]
Moved from https://github.com/gobuffalo/buffalo/issues/478

Since cockroachdb is written in Go, and speaks the postgres wire protocol, and is horizontally scalable, it would be a great match for Buffalo. When I manually created the three database (deal_development, deal_test, deal_production) in cockroach, everything worked just fine: "buffalo dev" started up just fine. 

Next I ran buffalo db drop -a, and while it appeared to succeed on the command line, from the cockroachdb admin http interface, those databases still exist. So drop does not work with cockroachdb.

Transcript:

~~~
jaten@jatens-MacBook-Pro-2 ~/go/src/github.com/gobuffalo/buffalo/deal (master) $ buffalo db drop -a
v3.23.2

DROP DATABASE deal_development;
DROP DATABASE deal_test;
DROP DATABASE deal_production;
jaten@jatens-MacBook-Pro-2 ~/go/src/github.com/gobuffalo/buffalo/deal (master) $ buffalo db create -a
v3.23.2

CREATE DATABASE deal_production;
Usage:
  buffalo db create [flags]

Flags:
  -a, --all   Creates all of the databases in the database.yml

Global Flags:
  -c, --config string   The configuration file you would like to use.
  -d, --debug           Use debug/verbose mode
  -e, --env string      The environment you want to run migrations against. Will use $GO_ENV if set. (default "development")
  -p, --path string     Path to the migrations folder (default "./migrations")

Error: couldn't create database deal_production: error creating PostgreSQL database deal_production: exit status 1

jaten@jatens-MacBook-Pro-2 ~/go/src/github.com/gobuffalo/buffalo/deal (master) $ 
~~~

I thought that perhaps the create failed because the tables already existed. But no, create fails even if the tables are not there.

~~~
jaten@jatens-MacBook-Pro-2 ~/go/src/github.com/gobuffalo/buffalo/deal (master) $ cockroach sql --insecure
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
^[[Jroot@:26257/> drop database deal_development;
^[[Jroot@:26257/> drop database deal_development;
DROP DATABASE
^[[Jroot@:26257/> drop database deal_test;
^[[Jroot@:26257/> drop database deal_test;
DROP DATABASE
^[[Jroot@:26257/> drop database deal_production;
^[[Jroot@:26257/> drop database deal_production;
DROP DATABASE
^[[Jroot@:26257/> ^D
^[[Jjaten@jatens-MacBook-Pro-2 ~/go/src/github.com/gobuffalo/buffalo/deal (master) $ buffalo db create -a
v3.23.2

CREATE DATABASE deal_development;
Usage:
  buffalo db create [flags]

Flags:
  -a, --all   Creates all of the databases in the database.yml

Global Flags:
  -c, --config string   The configuration file you would like to use.
  -d, --debug           Use debug/verbose mode
  -e, --env string      The environment you want to run migrations against. Will use $GO_ENV if set. (default "development")
  -p, --path string     Path to the migrations folder (default "./migrations")

Error: couldn't create database deal_development: error creating PostgreSQL database deal_development: exit status 1
$
~~~

Are there any logs about what went wrong? The error message here isn't very informative.

* * *

In any case, it would be great to be able to use cockroachdb with buffalo.

Since it is a Go program, cockroach is a single all-in-one binary, and it is very easy to get started with: 

https://www.cockroachlabs.com/docs/install-cockroachdb.html

**Comment by markbates** [comment]
@glycerine have you tried running with the `--debug` flag to see what is happening under the covers?

Maybe the problem is that it is using the Postgres command line tools under the covers, and those don't work with Cockroach?

I would definitely be open to having Cockroach support, but I suspect it's not quite as easy as one would think.

**Comment by glycerine** [comment]
Ah, its too bad the error messages from the `createdb` invocation aren't surfaced, but the --debug flag told me what to run:

~~~
$ createdb -e -h 127.0.0.1 -p 26257 -U postgres deal_development
CREATE DATABASE deal_development;
createdb: database creation failed: ERROR:  only root is allowed to CREATE DATABASE
~~~

This was an easy fix.

I changed the username to `root` in database.yml (in all three places), and now `buffalo db create -a` and `buffalo db drop -a` work fine with cockroach.

**Comment by glycerine** [comment]
Now that I know how to run the test suite for pop (running test.sh), its clear that #82 should stay open even with #91, as there are a few things that Cockroachdb doesn't handle the same way as postgres. In particular, the UUID type support appears to be different.

https://www.cockroachlabs.com/docs/stable/sql-faqs.html#does-cockroachdb-support-a-uuid-type says

> Does CockroachDB support a UUID type?
> Not at this time, but storing a 16-byte array in a BYTES column should perform just as well.

references: 

https://github.com/cockroachdb/cockroach/issues/14595

https://github.com/cockroachdb/cockroach/issues/8719

https://www.cockroachlabs.com/docs/stable/functions-and-operators.html   mentions: experimental_uuid_v4(), unique_rowid() , and uuid_v4() 

With database.yml specifying user: root
and cockroachdb started with "cockroach start --insecure -p 5432":

~~~
./test.sh
github.com/markbates/pop/soda
Testing postgres
./tsoda -v
v3.23.3

dropped database pop_test
v3.23.3

[POP]: Loading config file from /Users/jaten/go/src/github.com/markbates/pop/database.yml
[POP]: Loading config file from /Users/jaten/go/src/github.com/markbates/pop/database.yml
[POP] 2017/06/30 16:47:52 Create pop_test (postgres://root:postgres@localhost:5432/pop_test?sslmode=disable)
[POP] 2017/06/30 16:47:52 createdb -e -h localhost -p 5432 -U root pop_test
created database pop_test
v3.23.3

[POP]: Loading config file from /Users/jaten/go/src/github.com/markbates/pop/database.yml
[POP]: Loading config file from /Users/jaten/go/src/github.com/markbates/pop/database.yml
[POP] 2017/06/30 16:47:52 CREATE TABLE "schema_migration" (
"version" VARCHAR (255) NOT NULL
);
CREATE UNIQUE INDEX "version_idx" ON "schema_migration" (version);
[POP] 2017/06/30 16:47:52 SELECT count(*) as row_count FROM schema_migration AS schema_migration WHERE version = $1 | ["20160808213308"]
[POP] 2017/06/30 16:47:52 CREATE TABLE "users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"alive" boolean,
"birth_date" timestamp,
"bio" text,
"price" numeric DEFAULT '1.00',
"email" VARCHAR (50) NOT NULL DEFAULT 'foo@example.com',
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "good_friends" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"first_name" VARCHAR (255) NOT NULL,
"last_name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "not_validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "callbacks_users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"before_s" VARCHAR (255) NOT NULL,
"before_c" VARCHAR (255) NOT NULL,
"before_u" VARCHAR (255) NOT NULL,
"before_d" VARCHAR (255) NOT NULL,
"after_s" VARCHAR (255) NOT NULL,
"after_c" VARCHAR (255) NOT NULL,
"after_u" VARCHAR (255) NOT NULL,
"after_d" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "books" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"title" VARCHAR (255) NOT NULL,
"user_id" UUID NOT NULL,
"isbn" VARCHAR (50) NOT NULL,
"id" SERIAL PRIMARY KEY
);

0.0557 seconds
[POP] 2017/06/30 16:47:52 pg_dump -s --dbname=postgres://root:postgres@localhost:5432/pop_test?sslmode=disable
pg_dump: server version: 9.5.0; pg_dump version: 9.4.5
pg_dump: aborting because of server version mismatch
Error: error from called function: error inside of calling function: error executing 20160808213308_setup_tests.up.fizz, sql: CREATE TABLE "users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"alive" boolean,
"birth_date" timestamp,
"bio" text,
"price" numeric DEFAULT '1.00',
"email" VARCHAR (50) NOT NULL DEFAULT 'foo@example.com',
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "good_friends" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"first_name" VARCHAR (255) NOT NULL,
"last_name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "not_validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "callbacks_users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"before_s" VARCHAR (255) NOT NULL,
"before_c" VARCHAR (255) NOT NULL,
"before_u" VARCHAR (255) NOT NULL,
"before_d" VARCHAR (255) NOT NULL,
"after_s" VARCHAR (255) NOT NULL,
"after_c" VARCHAR (255) NOT NULL,
"after_u" VARCHAR (255) NOT NULL,
"after_d" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "books" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"title" VARCHAR (255) NOT NULL,
"user_id" UUID NOT NULL,
"isbn" VARCHAR (50) NOT NULL,
"id" SERIAL PRIMARY KEY
);: error inside of calling function: pq: syntax error at or near "uuid"
CREATE TABLE "users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"alive" boolean,
"birth_date" timestamp,
"bio" text,
"price" numeric DEFAULT '1.00',
"email" VARCHAR (50) NOT NULL DEFAULT 'foo@example.com',
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "good_friends" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"first_name" VARCHAR (255) NOT NULL,
"last_name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "not_validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "callbacks_users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"before_s" VARCHAR (255) NOT NULL,
"before_c" VARCHAR (255) NOT NULL,
"before_u" VARCHAR (255) NOT NULL,
"before_d" VARCHAR (255) NOT NULL,
"after_s" VARCHAR (255) NOT NULL,
"after_c" VARCHAR (255) NOT NULL,
"after_u" VARCHAR (255) NOT NULL,
"after_d" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "books" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"title" VARCHAR (255) NOT NULL,
"user_id" UUID NOT NULL,
          ^

Usage:
  soda migrate [flags]
  soda migrate [command]

Aliases:
  migrate, m

Available Commands:
  down        Apply one or more of the 'down' migrations.
  reset       The equivalent of running `migrate down` and then `migrate up`
  status      Displays the status of all migrations.
  up          Apply all of the 'up' migrations.

Flags:
  -h, --help   help for migrate

Global Flags:
  -c, --config string   The configuration file you would like to use.
  -d, --debug           Use debug/verbose mode
  -e, --env string      The environment you want to run migrations against. Will use $GO_ENV if set. (default "development")
  -p, --path string     Path to the migrations folder (default "./migrations")

Use "soda migrate [command] --help" for more information about a command.

error from called function: error inside of calling function: error executing 20160808213308_setup_tests.up.fizz, sql: CREATE TABLE "users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"alive" boolean,
"birth_date" timestamp,
"bio" text,
"price" numeric DEFAULT '1.00',
"email" VARCHAR (50) NOT NULL DEFAULT 'foo@example.com',
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "good_friends" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"first_name" VARCHAR (255) NOT NULL,
"last_name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "not_validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "callbacks_users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"before_s" VARCHAR (255) NOT NULL,
"before_c" VARCHAR (255) NOT NULL,
"before_u" VARCHAR (255) NOT NULL,
"before_d" VARCHAR (255) NOT NULL,
"after_s" VARCHAR (255) NOT NULL,
"after_c" VARCHAR (255) NOT NULL,
"after_u" VARCHAR (255) NOT NULL,
"after_d" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "books" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"title" VARCHAR (255) NOT NULL,
"user_id" UUID NOT NULL,
"isbn" VARCHAR (50) NOT NULL,
"id" SERIAL PRIMARY KEY
);: error inside of calling function: pq: syntax error at or near "uuid"
CREATE TABLE "users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"alive" boolean,
"birth_date" timestamp,
"bio" text,
"price" numeric DEFAULT '1.00',
"email" VARCHAR (50) NOT NULL DEFAULT 'foo@example.com',
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "good_friends" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"first_name" VARCHAR (255) NOT NULL,
"last_name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "not_validatable_cars" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"name" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "callbacks_users" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"before_s" VARCHAR (255) NOT NULL,
"before_c" VARCHAR (255) NOT NULL,
"before_u" VARCHAR (255) NOT NULL,
"before_d" VARCHAR (255) NOT NULL,
"after_s" VARCHAR (255) NOT NULL,
"after_c" VARCHAR (255) NOT NULL,
"after_u" VARCHAR (255) NOT NULL,
"after_d" VARCHAR (255) NOT NULL,
"id" SERIAL PRIMARY KEY
);
CREATE TABLE "books" (
"created_at" timestamp NOT NULL,
"updated_at" timestamp NOT NULL,
"title" VARCHAR (255) NOT NULL,
"user_id" UUID NOT NULL,
          ^


~~~

**Comment by markbates** [comment]
Yeah, it looks like we would definitely need a "proper" translator for Coachroach. I'm working on setting up a Dockerfile to make testing easier for contributors. It's probably the easiest way to have all of the correct databases installed and setup for everyone to run their tests against.

**Comment by glycerine** [comment]
Docker sounds like a great solution. If you don't already have a MySQL docker image, I created a pair of docker images for testing MySQL master/slave under GTIDs setup a couple months back; you could use the master image easily.

https://hub.docker.com/r/jaten/gtid56master/

**Comment by glycerine** [comment]
jordanlewis wrote on https://github.com/cockroachdb/cockroach/issues/16832
>  You'll be able to use UUIDs in the next major release of CockroachDB.

Once that comes out, re-evaluate and see where this ticket stands.

**Comment by markbates** [comment]
I think waiting until that release makes a lot of sense.

**Comment by mclark4386** [comment]
They are to at least 1.1.1 now... would be happy to help as much as I can. Where would you like me to start?

**Comment by markbates** [comment]
@mclark4386 I don't know much about cockroach, but I would start by copying all things Postgres and changing them to Cockroach, ie things like `postgresql.go`, `fizz/translators/postgres.go`, etc... then adding a cockroach setup to `docker-compose.yml`, edit `test.sh`, and then run, see what shakes out. that's pretty much what I would do. :)

**Comment by mclark4386** [comment]
@markbates will do! thanks!

**Comment by mclark4386** [comment]
Having to move back over to the frontend for a week or 3, so I can't complete this just this second... sorry. I'll be back on it after that. 
As an update, I got it to where I _think_ everything is working minus migrations that would involve altering a column that would cause a data migration (set not null, changing type, etc). I have plans to  also build out a generic schema lookup as well... but alas it will have to wait on more pressing matters^^;

here is what I have for cockroach so far: https://github.com/mclark4386/pop/tree/cockroach

**Comment by BigJoe01** [comment]
What happens with cockroach integration, current state?

**Comment by mclark4386** [comment]
Between the holidays, having to finish up this project and all the work I found that needed to be done I haven't had time to come back around... I may soon, but you are welcome to fork my work above or use it as reference if you need it sooner. And if you don't need column changing migrations you can use it now, I'm pretty sure.

**Comment by mclark4386** [comment]
implemented in #179 ^^
